Fixed #11964 -- Added support for check constraints in model Meta.

Ian-Foote opened the pull request:
https://code.djangoproject.com/ticket/11964

Ian-Foote commented on django/db/models/constraints.py:
```diff
@@ -0,0 +1,21 @@
+        quote_name = schema_editor.quote_name
+        connection = schema_editor.connection
+
+        sql, params = self.constraint.as_sql(compiler, connection)
```
From #8056, it looks like I can do:
```python
compiler = connection.ops.compiler('SQLCompiler')(None, connection, 'default')
```

Ian-Foote commented on django/db/models/constraints.py:
```diff
@@ -0,0 +1,22 @@
+    def create_sql(self, model, schema_editor, using=''):
+        quote_name = schema_editor.quote_name
+        connection = schema_editor.connection
+        compiler = connection.ops.compiler('SQLCompiler')(None, connection, 'default')
```
This gets me further. Unfortunately, I appear to need a more complete `query` instead of `None`, so `Col.as_sql` can be evaluated.

Ian-Foote commented on tests/migrations/test_operations.py:
```diff
@@ -1489,6 +1492,24 @@ def test_remove_index(self):
+        """Test the AddConstraint operation."""
+        project_state = self.set_up_test_model('test_addconstraint')
+
+        where = WhereNode([GreaterThan(Col('pink', models.F('pink')), 2)])  # Q(pink__gt=2)
```
Long term, I want to be able to use `Q(pink__gt=2)` here, but for now using this avoids the need to implement `Q.as_sql`.

Ian-Foote commented on tests/migrations/test_state.py:
```diff
@@ -1002,7 +1002,7 @@ class Meta:
         self.assertIs(author_state.fields[2][1].null, False)
         self.assertIs(author_state.fields[3][1].null, True)
-        self.assertEqual(author_state.options, {'swappable': 'TEST_SWAPPABLE_MODEL', 'indexes': []})
+        self.assertEqual(author_state.options, {'swappable': 'TEST_SWAPPABLE_MODEL', 'indexes': [], "constraints": []})
```
I haven't really thought whether this is the best way to fix these tests, but it is sufficient for now.

Ian-Foote commented:
I'm not sure where best to put MySQL error handling. It supports the syntax, but [doesn't actually do anything with it](https://dev.mysql.com/doc/refman/5.7/en/create-table.html#create-table-indexes-keys), so I think we should fail loud somewhere.

Ian-Foote commented:
I'm also not sure how best to handle SQLite. It [appears to support the feature](http://www.sqlite.org/syntax/table-constraint.html), but I get a syntax error from SQLite with the syntax that works on postgres. I think this is because that uses `ALTER TABLE`, which [SQLite only supports a limited form of](http://www.sqlite.org/lang_altertable.html), but I'm not clear what to do instead.

Django does set `supports_column_check_constraints` to `False` for SQLite's db features, so perhaps working around this limitation would be difficult.

Ian-Foote commented:
@jarshwah @MarkusH I'd appreciate your feedback on this. There's definitely some work left to do - docs, MySQL, SQLite, naming consistency - but I think the core is mostly there.

I'd also like to add a test to check an `IntegrityError` is actually raised if a `Check` constraint is violated, but I'm not sure how to set up the model to test so the `Check` is applied in the database.

auvipy commented on docs/ref/models/options.txt:
```diff
@@ -431,6 +431,26 @@ Django quotes column and table names behind the scenes.
+
+.. attribute:: Options.constraints
+
+    .. versionadded:: 1.11
```
should be 2.0

Ian-Foote commented:
This now just needs to do something sensible for MySQL and the rest is cleanup and edge-cases.

Ian-Foote commented on django/db/models/sql/query.py:
```diff
@@ -2108,3 +2108,435 @@ def update_join_types(self, query):
         return to_demote
+
+
+class ExpressionConstraintQuery:
```
It may be possible to simplify this further. I'm also not 100% certain I've thought of every edge-case, so I may have accidentally deleted something important.

adamchainz replied:
This *is* a lot of stuff to duplicate, especially as you note below that you don't understand everything. I'm concerned that future expression refactors/features will be missed in constraints and things will asplode.

Ian-Foote replied:
Yeah - that's a good point. I'll take another pass and see if I can minimise the amount of duplicated code.

aaugustin replied:
I'm out of my depth here; my instinct would be to figure out what the responsibilities of Q / Query / etc. are, which bits are needed for constraints, and refactor to share these bits.

Malcolm's "Dungeon Master's guide to Django's ORM" talk may help.

timgraham replied:
For example, `_force_pk` (duplicated below) was removed in ec50937bcbe160e658ef881021402e156beb0eaf.

Ian-Foote commented on django/db/models/sql/query.py:
```diff
@@ -2108,3 +2108,435 @@ def update_join_types(self, query):
+                target_clause.add(child_clause, connector)
+        return target_clause
+
+    def names_to_path(self, names, opts, allow_many=True, fail_on_missing=False):
```
I've not put any significant effort into understanding this method yet, so it may well be possible to simplify.

Ian-Foote commented on tests/queries/test_query.py:
```diff
@@ -0,0 +1,100 @@
+from .models import Author, Item, ObjectC, Ranking
+
+
+class TestExpressionConstraintQuery(TestCase):
```
Ideas for more tests to add here would be appreciated.

atombrella commented on django/db/models/sql/query.py:
```diff
@@ -2108,3 +2108,435 @@ def update_join_types(self, query):
         return to_demote
+
+
+class ExpressionConstraintQuery:
```
It feels odd why so much code needs to be duplicated. Have you considered letting `Query` subclass this?

Ian-Foote replied:
I hadn't considered that. My main focus has been trying to get something that works well for this use case. Refactoring to allow code re-use is a secondary concern.

auvipy commented:
I have updated the ticket to be visible on the review-queue

adamchainz commented on django/db/backends/mysql/features.py:
```diff
@@ -26,6 +26,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     can_release_savepoints = True
     atomic_transactions = False
     supports_column_check_constraints = False
+    supports_table_check_constraints = False
```
Ah, so MariaDB *does* support them... I guess this'll help push me to get first class MariaDB support into Django after this gets merged.

Also there are some wonderful hacks to make them work as blogged by the MySQL team here: http://mysqlserverteam.com/new-and-old-ways-to-emulate-check-constraints-domain/ . I think the generated column method could potentially be used by Django, it means making very different SQL on MySQL though. Not sure if it's worth it, at least initially.

Ian-Foote replied:
I also think it's not really worth it.

adamchainz commented on django/db/models/base.py:
```diff
@@ -1655,13 +1657,29 @@ def _check_long_column_names(cls):
+    @classmethod
+    def _check_constraints(cls):
+        errors = []
+        if connection.features.supports_table_check_constraints:
```
I was confused by this briefly, I think you should change the below to be

```
if (
    not connection.features.supports_table_check_constraints and
    any(isinst...
):
    errors.append(
```

timgraham replied:
This should use the same logic in `_check_long_column_names()` rather than only examining the default database connection.

adamchainz commented on django/db/models/base.py:
```diff
@@ -1655,13 +1657,29 @@ def _check_long_column_names(cls):
+        if any(isinstance(constraint, CheckConstraint) for constraint in cls._meta.constraints):
+            errors.append(
+                checks.Warning(
+                    'Check constraints are not enforced on MySQL.',
```
come on, no need to pick on MySQL... "Check constraints are not supported on your database backend"

timgraham replied:
You can use `connection.display_name` to get the database's name.

adamchainz commented on django/db/models/base.py:
```diff
@@ -1655,13 +1657,29 @@ def _check_long_column_names(cls):
 
 # ORDERING METHODS #########################
 
+
```
is this necessary?

Ian-Foote replied:
Flake8 was complaining about this.

adamchainz commented on django/db/models/constraints.py:
```diff
@@ -0,0 +1,54 @@
+    def __eq__(self, other):
+        if not isinstance(other, CheckConstraint):
+            return NotImplemented
+        return self.name == other.name and self.constraint == other.constraint
```
I think normally we don't `raise NotImplemented` but just return `False` so

```py
return (
    isinstance(other, CheckConstraint) and
    self.name == other.name and
    self.constraint == other.constraint
)
```

adamchainz commented on docs/ref/models/check-constraints.txt:
```diff
@@ -0,0 +1,43 @@
+``CheckConstraint`` options
+===========================
+
+.. class:: CheckConstraint(constraint, name)
```
I think it would be more natural if `name` was first, especially because `Q` objects can be arbitrarily complex.

aaugustin replied:
Do constraints really need an explicit, user-defined name?

Django autogenerates a name for indexes. I would expect it do the same for constraints. This suggests `CheckConstraint(constraint, name=None)`.

aaugustin replied:
(This could be retrofitted later.)

adamchainz replied:
That's a good point. The same algorithm can probably be used.

MarkusH replied:
Even if indexes use autogenerated names when no explicit ones are given, the mere presence of maintaining an autogenerated name even though the source for the generator (field names) can change is a pain. We can add autogeneration at some point, but taking it away is not really possible without too much trouble for our users.

The name attribute is last for `Index` as well. I think I'd leave it there.

timgraham commented:
buildbot, test on oracle.

timgraham commented:
buildbot, test on oracle.

aaugustin commented:
This needs a changelog entry.

aaugustin commented on django/db/models/expressions.py:
```diff
@@ -713,6 +713,33 @@ def get_db_converters(self, connection):
+
+    def as_sql(self, compiler, connection):
+        qn = compiler.quote_name_unless_alias
+        return "%s" % (qn(self.target.column)), []
```
`"%s" % (qn(self.target.column))` is just `qn(self.target.column)`, isn't it?

aaugustin commented:
This is shaping up really well. Thanks Ian!

The most significant sticking point is the code duplication in `query.py`. Do you think you can figure it out? I was never very familiar with that part of the ORM myself and I'm unlikely to be able to help :-(

I brought up a question about the necessity of user-defined names above, which I think is important, but I won't block adding the feature for this reason. A follow-up ticket would do.

timgraham commented on django/db/backends/base/schema.py:
```diff
@@ -327,6 +333,18 @@ def remove_index(self, model, index):
 
+    def add_constraint(self, model, constraint):
+        """
+        Add a check constraint to a model.
```
Use a single line docstring.

timgraham commented on django/db/models/expressions.py:
```diff
@@ -713,6 +713,33 @@ def get_db_converters(self, connection):
+
+    def __repr__(self):
+        return "{}({})".format(
+            self.__class__.__name__, self.target)
```
Put this on the previous line.

timgraham commented on docs/ref/checks.txt:
```diff
@@ -310,6 +310,7 @@ Models
   underscore as it collides with the query lookup syntax.
 * **models.E024**: The model name ``<model>`` cannot contain double underscores
   as it collides with the query lookup syntax.
+* **models.W025**: :class:`~django.db.models.CheckConstraint` constraints are not enforced by MySQL.
```
Wrap documentation at 79 characters please.

timgraham commented on docs/ref/models/check-constraints.txt:
```diff
@@ -0,0 +1,43 @@
+
+.. versionadded:: 2.0
+
+The CheckConstraint class allows creating database check constraints. They can be added using the
```
Add double backticks around CheckConstraint.

atombrella commented on django/db/migrations/operations/models.py:
```diff
@@ -834,3 +834,88 @@ def deconstruct(self):
+        )
+
+    def describe(self):
+        return 'Remove constraint %s from model %s' % (
```
Wrap `%s` inside single-quotes?

Ian-Foote commented on django/db/models/constraints.py:
```diff
@@ -0,0 +1,54 @@
+            return NotImplemented
+        return self.name == other.name and self.constraint == other.constraint
+
+    def deconstruct(self):
```
I'm not clear what you're suggesting here with the `path`.

Ian-Foote commented on tests/migrations/test_operations.py:
```diff
@@ -1572,6 +1616,89 @@ def test_alter_index_together_remove(self):
+
+    @skipUnlessDBFeature('supports_table_check_constraints')
+    def test_remove_constraint(self):
+        """
```
It matches similar tests. E.g. `test_remove_index`.

MarkusH commented on docs/ref/models/options.txt:
```diff
@@ -455,6 +455,26 @@ Django quotes column and table names behind the scenes.
+
+.. attribute:: Options.constraints
+
+    .. versionadded:: 2.0
```
2.1

MarkusH commented on docs/ref/models/check-constraints.txt:
```diff
@@ -0,0 +1,43 @@
+
+.. currentmodule:: django.db.models
+
+.. versionadded:: 2.0
```
2.1

MarkusH commented on django/db/migrations/autodetector.py:
```diff
@@ -175,8 +176,10 @@ def _detect_changes(self, convert_apps=None, graph=None):
+        self.create_altered_constraints()
         # Generate index removal operations before field is removed
         self.generate_removed_indexes()
+        self.generate_removed_constraints()
```
In the context of e.g. unique constraints depending on indexes to maybe remove constraints before removing indexes?

jackton1 commented on django/db/models/sql/query.py:
```diff
@@ -993,15 +999,22 @@ def resolve_expression(self, query, *args, **kwargs):
+    def resolve_lookup_value(self, value, can_reuse, allow_joins, simple_col):
         if hasattr(value, 'resolve_expression'):
-            value = value.resolve_expression(self, reuse=can_reuse, allow_joins=allow_joins)
+            if isinstance(value, F):
```
```python
kwargs = {'reuse': can_reuse, 'allow_joins': allow_joins}
if isinstance(value, F):
    kwargs['simple_col'] = simple_col
value = value.resolve_expression(self, **kwargs)
```

carltongibson commented:
Issue: https://code.djangoproject.com/ticket/11964
PR: https://github.com/django/django/pull/7615

---

Hi @Ian-Foote. Thanks for the extended effort here.

Reviewing the history there are just a few small points outstanding:

- We'll have to target 2.2 now, so those references need updating. 
- Via @aaugustin: "This needs a changelog entry."
- Via @timgraham: "The migration operations should be documented in docs/ref/migration-operations.txt."

I pulled in your test for the `IntegrityError` [here](https://github.com/carltongibson/django/commit/710443ad00231b883d98737cd072eb3caffaca6a). That already behaves as expected. 

I was going to say, can we make the MySQL `add_constraint()`/`remove_constraint()` no-ops, with an extra warning maybe? But the test failed **without crashing** when run on MySQL. I then noticed your comment: 

> [MySQL] supports the syntax, but doesn't actually do anything with it...

Thus I think we're good there. 

I would push the test straight here but I didn't understand this: 

> ...I'm not sure how to set up the model to test so the Check is applied in the database.

Perhaps that's not relevant any more? 

Other than that, this looks super!

Ian-Foote commented:
Oh that's annoying - I didn't notice the 2.1 cut-off approaching or I'd have flagged this a bit earlier.

> ...I'm not sure how to set up the model to test so the Check is applied in the database.
I don't remember where I said that, so I'm guessing it's not relevant any more.

I'll make sure the outstanding points about documentation are resolved at the DjangoCon Europe sprints.

Ian-Foote commented on django/db/backends/base/schema.py:
```diff
@@ -302,7 +308,7 @@ def create_model(self, model):
         sql = self.sql_create_table % {
             "table": self.quote_name(model._meta.db_table),
-            "definition": ", ".join(column_sqls)
+            "definition": ", ".join(chain(column_sqls, constraints))
```
We probably can, but I don't think it's better.

adamchainz replied:
but it's faster because there's no function call overhead

```
In [1]: from itertools import chain

In [2]: a = ['1', '2']

In [3]: b = ['buckle', 'my', 'shoe']

In [4]: %timeit list(chain(a, b))
506 ns ± 2.19 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit list((*a, *b))
313 ns ± 4.46 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Ian-Foote replied:
Ok - I'll do it then.

Ian-Foote commented on django/db/models/constraints.py:
```diff
@@ -0,0 +1,54 @@
+        return "<%s: constraint='%s' name='%s'>" % (self.__class__.__name__, self.constraint, self.name)
+
+    def __eq__(self, other):
+        return (isinstance(other, CheckConstraint) and
```
I'm not quite clear what style you're recommending here - can you be more explicit?

adamchainz replied:
```python
hanging = (
    indentation,
    has,
    a,
    newline,
    after,
    opening,
    bracket,
)
```

Ian-Foote commented:
@carltongibson I think I've resolved the outstanding points now. I've included the test you linked, added the missing docs and fixed @felixxm's comments.

carltongibson commented:
OK. Great stuff @Ian-Foote! I will have a look again this week.

carltongibson commented:
Yes. Just to emphasis, I totally agree: this is great. My queries are merely _clarificatory_.

carltongibson commented:
Hey @Ian-Foote. 

Thanks for that. Super. I think, for me, if you could add a docstring to `SimpleCol` that would be handy. 

Beyond that, lets call it RFC. Good work. 

(Now can we just get forms to parse constraints for validation there...? 🙂)

Ian-Foote commented:
@carltongibson Docstring added.

adamchainz commented:
🎉 Thanks for your great work and persistence, @Ian-Foote and everyone else 🎉
